NUT-XX: Proof of Liabilities (PoL) (#388)
cashubtc/nuts: pol.md

## Architecture Overview

A Cashu mint acts as a custodian of backing assets. Outstanding liabilities are proven using two synchronized 256-depth Sparse Merkle Sum Trees (MS-SMT):

1. **Issued Tree (Promises):** Tracks all signed blinded messages $B'$.
2. **Spent Tree (Proofs Used):** Tracks all spent proof secrets $Y = \text{hash\_to\_curve}(\text{secret})$.

This prevents the mint from omitting or manipulating liabilities. Epoch-based state roots are periodically committed to the Bitcoin blockchain via OpenTimestamps, establishing an immutable history.

---

## MS-SMT Specifications

The MS-SMT has a fixed depth of 256 levels (level 0 at leaves, 256 at root).

### 1. Leaf Index and Keys

- **Issued Leaf Index ($I_{\text{issued}}$):**
  - Index: $\text{SHA256}(\text{bytes}(B'))$ (using the 33-byte compressed public key representation of $B'$) parsed as a big-endian integer.
  - Value: Face amount of the signed promise.
- **Spent Leaf Index ($I_{\text{spent}}$):**
  - Index: $\text{SHA256}(\text{bytes}(Y))$ (using the 33-byte compressed public key representation of $Y$) parsed as a big-endian integer.
  - Value: Face amount of the spent proof.

### 2. Node Structure & Hashing

> Each node is represented as $(hash, sum\_value)$ where $hash$ is 32 bytes and $sum\_value$ is an 8-byte big-endian integer.

Reviewer: We should be explicit about overflow and duplication behavior. e.g.

```md
If `sum_L + sum_R >= 2^64`, tree construction MUST fail.

A mint MUST NOT insert the same issued item `B'` more than once in the same issued tree, and MUST NOT insert the same spent item `Y` more than once in the same spent tree. Duplicate insertion MUST be treated as an implementation error.
```

Author: > A mint MUST NOT insert the same issued item `B'` more than once in the same issued tree, and MUST NOT insert the same spent item `Y` more than once in the same spent tree. Duplicate insertion MUST be treated as an implementation error.

It's not physically possible to insert duplicates  because it's the same entry.

---

## Epoch Manifests & On-Chain Commitments

Every epoch interval (e.g., 24 hours), the mint constructs and signs an Epoch Manifest:

1. **Sort Keysets:** Sort all unexpired keyset IDs alphabetically.
2. **Commitment Data:** Concatenate the UTF-8 `keyset_id`, 32-byte `root_issued_hash`, and 32-byte `root_spent_hash` for each keyset sequentially.
3. **Global Digest:** Compute $\text{SHA256}(\text{commitment\_data})$.
4. **OTS Submission:** Submit this digest to OTS calendar servers to obtain a binary receipt.
5. **Manifest Message:** Construct a colon-separated UTF-8 string:
   `"{keyset_id}:{epoch_index}:{timestamp}:{root_issued_hash}:{root_issued_sum}:{root_spent_hash}:{root_spent_sum}:{outstanding_balance}:{ots_receipt}"`
   where `ots_receipt` is the hex-encoded OTS receipt.
6. **Signing:** Sign the message with a BIP-340 Schnorr signature using the mint's master NUT-06 private key signing the SHA256 digest of this serialized manifest string.
7. **Publish:** Store and publish the signed manifest, signatures, and OTS receipts.

---

## Compact Bitmasked Sibling Proofs

> To minimize proof size, default empty sibling nodes are omitted. Instead, the mint returns a 256-bit bitmask where the $d$-th bit indicates if the sibling at level $d$ is non-empty ($1$) or empty ($0$, to be replaced by the precomputed default empty node).

Reviewer: For clarity

```suggestion
To minimize proof size, default empty sibling nodes are omitted. The mint returns a 256-bit `compact_mask`, serialized as a 32-byte big-endian integer and displayed as 64 lowercase hex characters.

Bit `d` of the integer, where `d = 0` is the leaf level and `d = 255` is the sibling below the root, indicates whether the sibling at level `d` is non-empty:

- `1`: the sibling is present in `siblings`.
- `0`: the verifier MUST use the precomputed default empty node for level `d`.
```

For detailed examples and test vectors, see the [test vectors][tests].

---

## HTTP API Specifications

### 1. Get Keyset Manifest

`GET /v1/pol/{keyset_id}/manifest`

- **Query Params:** `epoch_index` (optional, integer)
- **Response:**

```json
{
  "keyset_id": "009a6154b71113b7",
  "epoch_index": 1,
  "timestamp": "2026-06-11T12:00:00Z",

>   "signing_pubkey": "020...",

Reviewer: Looks like a compressed pubkey, not a 32-byte BIP340

- **Query Params:** `epoch_index` (optional, integer)
- **Request Body:**

```json
{ "blinded_messages": ["02b1a..."] }
```

- **Response:**

```json
{
  "proofs": [
    {
      "item": "02b1a...",
      "index": "8a31...",
      "value": 1000,
      "compact_mask": "0x301a...",
      "siblings": [{ "hash": "b4a1...", "sum": 500 }]
    }
  ]
}
```

### 3. Query Spent Tree Proofs

`POST /v1/pol/{keyset_id}/proofs/spent`

- **Query Params:** `epoch_index` (optional, integer)
- **Request Body:**

```json
{ "ys": ["02b1a..."] }
```

- **Response:** Same format as `/proofs/issued` with the $Y$ point hex string in the `item` field.

---

## Signed Transactional Proof of Liability Receipts

The mint **MUST** return a cryptographically signed **PoL Receipt** nested inside each input spent and output returned during state-transitioning actions (`mint`, `melt`, and `swap`).

### 1. Receipt JSON Schema

```json
{
  "target_epoch": 12,
  "signature": "<hex_encoded_signature>"
}
```

### 2. Message Formats and Cryptography

Each receipt is signed under the keyset's per-amount private key (`private_keys[amount]`) corresponding to the note's denomination.

> - **Message to Sign:** `<point_hex>:<target_epoch_decimal_string>`

Reviewer: Should this commit to more that just these two?

Author: I think no

| :-------- | :-------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `00`/`01` | secp256k1 | BIP340 Schnorr over $\text{SHA256}(\text{message})$. Verified against `public_keys[amount]`.                                                                                                                                                                                   |
| `02`      | BLS12-381 | $\sigma = a \cdot H_{G1}(\text{message})$ (compressed G1). Verified via $e(\sigma, G2) == e(H_{G1}(\text{message}), K)$ where $K = \text{public\_keys}[amount]$ (G2). $H_{G1}$ is RFC 9380 `hash_to_curve_G1` under DST `"Cashu_PoL_Receipt_BLS12381G1_XMD:SHA-256_SSWU_RO_"`. |

### 3. Response Alignment

Returned receipts are fully order-preserving (1:1 index matching of request inputs/outputs):

> - **Mint (`POST /v1/mint/bolt11` & `/v1/mint/bolt11/batch`):** Nested in `pol_receipt` of each `BlindSignature` inside `signatures`.

Reviewer: should use the generic-method language instead `/v1/mint/{method}`

### Step 1: Verify Manifest Signature

Verify the BIP-340 Schnorr signature `mint_signature` against the mint's master public key (`signing_pubkey` from `/v1/info`) over the SHA256 digest of the constructed epoch string.

### Step 2: Validate OpenTimestamps Attestation

1. **Upgrade Receipt:** Post `ots_receipt` to a calendar upgrade endpoint (e.g., `https://alice.btc.calendar.opentimestamps.org/upgrade`) to fetch the Merkle path.
2. **Scan for Block:** Find block header attestation tag `0x00 0x05` (`A_BLOCKHEADER`). If pending tag `0x00 0x06` is found, the receipt is still awaiting block confirmation.
3. **Parse Height:** Decode the Bitcoin block height (serialized as a VarInt) immediately following the tag.
4. **Confirm:** Check via an independent explorer (e.g., `https://mempool.space/api/...`) that the block height exists, matches the manifest timestamp, and has sufficient confirmations.

### Step 3: Validate Issued Path Walks

For each held active token:

1. Reconstruct $B'$ (BDHKE: $B' = Y + rG$; BLS: $B' = r \cdot Y$).
2. Leaf index $I = \text{SHA256}(B')$ parsed as a big-endian integer.
3. Walk up the 256 levels:
   - For $d \in [0, 255]$:

>      - The $d$-th bit of the bitmask determines child position.

Reviewer: This is inconsistent with the earlier paragraph where the $d$-th bit of the bitmask determines supplied vs. default node, not child position.

#### Default Empty Nodes
To represent empty branches without storing 2^256 nodes, default empty node values are precomputed for each level `d` from 0 to 256:
- **At level 0 (leaf):**
  - `hash_0` is the SHA256 hash of empty bytes (`b""`).
  - `sum_0` is `0`.
- **At level d (for d >= 1):**
  - `sum_d` is `0` (the sum of both default child sums: `0 + 0`).
  - `hash_d` is the SHA256 hash of the concatenated bytes:
    `hash_{d-1} + hash_{d-1} + bytes_8(sum_{d-1}) + bytes_8(sum_{d-1})`
    where `bytes_8(x)` is the 8-byte big-endian representation of integer `x`.

#### Parent Node Computation
For any level `d` from 0 to 255, two neighboring sibling nodes `L = (hash_L, sum_L)` and `R = (hash_R, sum_R)` at level `d` are aggregated into parent node `P = (hash_P, sum_P)` at level `d+1`:
- `sum_P` is the integer sum: `sum_L + sum_R`
- `hash_P` is the SHA256 hash of the concatenated bytes:
  `hash_L + hash_R + bytes_8(sum_L) + bytes_8(sum_R)`

---

## Epoch Manifest Creation and On-Chain Commitments

To prevent the mint from rewriting historic trees or showing different root values to different users, roots are committed to epochs synchronized across the entire mint.

> Every epoch interval (e.g., 24 hours), the mint builds the Issued and Spent trees for all unexpired keysets, publishes on-chain commitments, and signs the resulting manifests:

Reviewer 2: What would be the plan for expired keysets? These can still have liabilities attached.

Author: Once the keyset expires, the mint simply stops publishing manifests including the roots for that keyset. The fact that they can still have liabilities is inconsequential IMO

  where `{B_hex}` is the hex-encoded string of the output's BlindedMessage `B'`, and `{target_epoch}` is the target upcoming epoch index string.
- **Signing key:** The keyset private key corresponding to the output's specific amount (`keyset.private_keys[amount]`).
- **Verification:** Wallets verify the BIP340 Schnorr signature on the SHA-256 hash of the formatted message against the keyset public key corresponding to the output's amount (`keyset.public_keys[amount]`).

#### B. Spent Inputs (Ecash Proofs)
For each spent proof in a swap or melt, the mint returns a list of `spent_receipts` in the response.
- **Message format to sign:**
  `{Y_hex}:{target_epoch}`
  where `{Y_hex}` is the hex-encoded public key string of the spent proof's curve point `Y = hash_to_curve(secret)`, and `{target_epoch}` is the target upcoming epoch index string.
- **Signing key:** The keyset private key corresponding to the spent input's specific amount (`keyset.private_keys[amount]`).
- **Verification:** Wallets verify the BIP340 Schnorr signature on the SHA-256 hash of the formatted message against the keyset public key corresponding to the spent input's amount (`keyset.public_keys[amount]`).

> 

Reviewer 2: The per-amount public keys in BLS12-381 (02) keysets (#371) are in G2, so a BIP340 receipt can't verify against public_keys[amount] as written.

To preserve the per-amount binding, perhaps the receipt signature could track the keyset's curve (BLS signature for 02)?

Otherwise, binding a receipt to its amount would need a separate secp key per amount for BLS keysets.

```suggestion
### 2. Message Formats and Keys

Each receipt is signed under the keyset's per-amount key (`private_keys[amount]`), binding it to the note's denomination. The signature scheme follows the keyset's curve (NUT-02 version byte):

| Version | Curve | Signature |
|-----------|-----------|------------------------------------------------------------|
| `00`/`01` | secp256k1 | BIP340 Schnorr over `SHA-256(message)` |
| `02` | BLS12-381 | `σ = a · H_G1(message)`, verified `e(σ, G2) == e(H_G1(message), K)` |

**Message to sign:** the UTF-8 bytes of the point hex, a literal `:`, and the target epoch index as a decimal string.
- **Output:** `<B'_hex>:<target_epoch>`, e.g. `02b1a...d4f:12`
- **Spent input:** `<Y_hex>:<target_epoch>`, e.g. `02e9c...a07:12`, where `Y = hash_to_curve(secret)`.

**secp256k1 (`00`/`01`):** verify the BIP340 signature against `public_keys[amount]`.

**BLS12-381 (`02`):** `H_G1` is `hash_to_curve_G1` (RFC 9380) under DST `"Cashu_PoL_Receipt_BLS12381G1_XMD:SHA-256_SSWU_RO_"`, which MUST differ from the BDHKE secret DST. `σ` is the compressed G1 encoding; `K = public_keys[amount]` (G2). Receipts MAY be batch-verified per NUT-00.

```

1. **Upgrade Proof (Optional):** Submit the binary `ots_receipt` to a public calendar server upgrade endpoint (e.g., `https://alice.btc.calendar.opentimestamps.org/upgrade`). If a mined block contains the commitment, the calendar returns an upgraded receipt containing the on-chain Merkle path.
2. **Scan for Block Attestation:** Scan the binary upgraded receipt for the Bitcoin Block Header attestation tag (`0x00` followed by `0x05` for `A_BLOCKHEADER`).
3. **Parse Block Height:** Parse the Bitcoin block height immediately following the tag (serialized as a standard variable-length VarInt).
4. **Verify Block Confirmation:** Query any independent public Bitcoin block explorer API (e.g. `https://mempool.space/api/blocks/tip/height`) to check that the parsed block height exists, matches the manifest timestamp, and has sufficient confirmations. If present, the attestation is fully confirmed. If the pending tag `0x00 0x06` is found instead, the receipt has been submitted to the calendar but is still awaiting blockchain anchoring.

### Step 3: Validate Issued Path Walks
For each active token held:

> 1. Reconstruct `B'` (if only holding `C` and `r`, compute `B' = Y + rG`).

Reviewer 2: Would need a BLS counterpart adding too (`B_ = r·Y`)

2. Compute the leaf index by parsing the SHA256 hash of `B'` as a big-endian integer.
3. Reconstruct the root hash and sum by walking up the 256 levels:
   - For `d` from 0 to 255:
     - The `d`-th bit of the bitmask determines whether the current node is the left or right child.
     - If the `d`-th bit of the compact mask is `1`, pop the next sibling from the response `siblings` array.
     - If `0`, use the precomputed default empty node `(default_hash[d], 0)`.
     - Hash parent and sum to get the node at level `d+1`.
4. Ensure the computed root matches `root_issued`.

### Step 4: Validate Spent Path Walks
For all spent tokens (history):
1. Compute `Y = hash_to_curve(secret)`.
2. Compute the leaf index by parsing the SHA256 hash of `Y` as a big-endian integer.
3. Perform the same walk up to verify inclusion in the `root_spent` tree.

### Step 5: Verify Solvency Equation
Check that the outstanding balance reported matches:
`outstanding_balance == root_issued_sum - root_spent_sum`

---

## Cryptographic Fraud Challenge (JSON format)

> If any verification path walk fails (e.g., the leaf is not present or has a sum value different from the token's face value, or the computed root doesn't match the signed root), the wallet generates a **Fraud Challenge**. This is an aggregated, self-contained JSON document that can be shared publicly to mathematically prove the mint committed perjury:

Reviewer 2: An honest mint might only get the note into an epoch after the target, for example a swap near the cutoff settles into N+1 instead of N. Since the trees are cumulative, checking the latest published epoch (≥ target_epoch) still finds it, whereas pinning the check to exactly `target_epoch` could falsely flag the mint. A note that never appears is still caught.

```suggestion
If any verification path walk fails in the latest published epoch `≥ target_epoch` (e.g., the leaf is not present or has a sum value different from the token's face value, or the computed root doesn't match the signed root), the wallet generates a **Fraud Challenge**. This is an aggregated, self-contained JSON document that can be shared publicly to mathematically prove the mint committed perjury:
```

Author: > whereas pinning the check to exactly target_epoch could falsely flag the mint

It's the mint that decides which epoch to commit them in, and lets the client know with a `pol_receipt`. So -in other words- it's the Mint's responsibility to make sure the entries appear by the specified epoch in the `pol_receipt`.

### Step 3: Validate Issued Path Walks

For each held active token:

1. Reconstruct the unsigned blinded message `B_` (BDHKE: `B_ = Y + rG`; BLS: `B_ = r * Y`).
2. Leaf index `I = SHA256(bytes(B_))` parsed as a big-endian integer.
3. Walk up the 256 levels:
   - For `d in [0, 255]`:
     - If the `d`-th bit is `1`, pop the next sibling from the proof's `siblings`.
     - If `0`, use default empty node `(default_hash[d], 0)`.
     - Compute the parent node `(hash, sum)` at level `d+1`.
4. Ensure the final root matches `root_issued`.

### Step 4: Validate Spent Path Walks

For spent tokens (history):

1. Compute `Y = hash_to_curve(secret)`.
2. Leaf index `I = SHA256(bytes(Y))` parsed as a big-endian integer.
3. Walk up the tree and verify matching `root_spent`.

### Step 5: Verify Liabilities Equation

Ensure:

```
outstanding_balance == root_issued_sum - root_spent_sum

> ```

Reviewer 3: This step is a consistency check of the manifest rather than a proof. All three values here come from the mint itself. The difference isn't a trustworthy liability figure even when checked: the spent sum can be inflated with fabricated leaves that no wallet can dispute. `outstanding_balance` is effectively a floor set by auditing users but not a verified total. 

I'm not sure this can be avoided. If not, limitations such as this could be stated in a dedicated section.

Since the field is fully derivable from `root_issued_sum - root_spent_sum`, it could be dropped from the manifest (and the signed epoch string) entirely which also eliminates the need for this step.

Author: > inflated with fabricated leaves that no wallet can dispute

Read the original pol scheme by @callebtc on how the mint is eventually caught doing this

Reviewer 3: Thanks for the pointer. How the mint gets caught in the PoL scheme may be worth working into the spec. The catch is collective: "only when multiple users come together and prove that the aggregate of their balances is larger than the outstanding balance of the report, they can prove that the mint has added fake entries to this list." A single wallet only knows its own tokens, so alone it can only catch edge cases involving it's own balance out-sizing the reported outstanding balance:

Current spec hands the check to individual wallets ("Wallets periodically audit their held and spent tokens"). So Step 5 as written is a check that no single actor can meaningfully act on. The group version that functionally catches a cheating mint lacks a mechanism so far. Three pieces could be useful here:

1. **A format wallets can pool.** The Fraud Challenge is close, but it can't just be collected and summed: it carries one Merkle path where a pooled entry needs two ("counted as issued" and "not counted as spent"), it's built to say "this check failed" while a pooled entry says the opposite ("my token checks out fine, and yet the totals don't add up"). Perhaps a rule for deduping and summing entries against a manifest is appropriate too.

2. **Define what `/proofs/spent` returns when a token is *not* in the tree.** A pooled entry has to prove its token was never redeemed otherwise the mint can dismiss the whole sum with "those are already among the spends I counted." The natural proof is the mint showing the token's slot is empty: the same sibling walk as an inclusion proof, but starting from the default empty leaf and landing on the anchored `root_spent`. The ingredients already exist in the doc (the test vectors precompute the empty nodes, the Fraud Challenge carries `actual_value: 0`) only the endpoint's response for absent items is left undefined.

3. **Keyset lifecycle in the manifests.** The group check becomes increasingly effective if a keyset's outstanding balance is forced to shrink over time through regular rotations.

Question: is the intent that wallets run Step 5 individually (and if so, what should a single wallet do with the result?), or that auditors aggregate across wallets? If aggregation is the model, should this NUT define the format they'd collect?

Author: > "only when multiple users come together and prove that the aggregate of their balances is larger than the outstanding balance of the report, they can prove that the mint has added fake entries to this list."

No, the users keep the root of the merkle-sum trees honest through random audits. Then, if the mint inflates (fill the spent book with fake ecash notes), the merkle-sum tree for the spent ecash will at some point be larger than the issued one, because when the keyset is rotated out it can no longer be minted from and therefore the issued book is pinned (only the spent book can change).

Author: > Question: is the intent that wallets run Step 5 individually (and if so, what should a single wallet do with the result?), or that auditors aggregate across wallets? If aggregation is the model, should this NUT define the format they'd collect?

Yes the intent is very much that. I left undefined how one would go to formally challenge the Mint to provide a proof of inclusion, but the idea is that the wallet claiming fraud "publishes" this challenge out in the open somewhere and other wallets can query the same Mint for an inclusion proof in that epoch. If it fails to deliver then it proves fraud.

Reviewer 3: Agreed on the mechanism here for the independent wallet audit. It's the spent tree growing against the locked issued tree on an inactive keyset that leads to revealing the fraud to any single auditor.

Noting two undefined mechanisms for consideration that could increase wallet audit effectiveness:

1. **Committing to a keyset rotation.** Rotations aren't currently explicit. Is it sufficient that a keyset beyond a certain age becomes itself a red flag to the auditor? A keyset's age is provable from its first appearance in the anchored manifest history.

2. **The spent tree could be verified as append-only.** Two anchored roots don't show leaf *removals* between epochs, so a mint could delete fake spends (as real ones are added) on an inactive keyset undetected. Publishing each epoch's added leaves (old root + additions = new root) would close that. The cost ofc is an additional, potentially large, payload.

Author: Point 2 is actually kind of critical. We need to rethink this. I propose we use a MMR (Merkle Mountain Range) instead of simple merkle-sum trees. The MMR is append-only and grows over time and we could capture a snapshot of it at every epoch, and commit it onchain with OTS. The proofs of inclusion the clients receive would be valid for that epoch, and further epochs would only require an "extension" thereof. The only downside is that with a MMR we cannot offer proofs of non-inclusion (but we don't really use them in the scheme).

Author: Addressed by the append-only sum-MMR redesign and the dedicated spent-side inflation/keyset rotation section. Fabricated spent leaves are now permanently committed and become detectable as inactive keysets wind down.

     - Push `(P_node, P_height)` back onto the `peaks` stack.
3. After all leaves are appended, the remaining nodes in the `peaks` stack (ordered from left to right) represent the current forest of MMR peaks.

### 4. Append-Only Consistency Verification

To verify that MMR M (of size m) is a valid append-only extension of MMR N (of size n, where m > n), an auditor compares their peak lists. This append-only property can be verified in two ways:

1. **Standard MMR Consistency Proofs:** Cryptographically proving that the peaks of the old MMR N are deterministically preserved and folded into the peaks of the new MMR M.
2. **Distributed Sibling Path Prefix Checks:** Any client holding a valid inclusion proof (such as for active or spent ecash) can re-request a new proof in the following epoch and verify that the old sibling path is preserved as a direct prefix of the new sibling path. If the mint modified or removed any historical entries, the internal hashes would change, immediately violating prefix preservation.

---

## Epoch Manifests & On-Chain Commitments

Every epoch interval (e.g., 24 hours), the mint constructs and signs an Epoch Manifest:

> 1. **Sort Keysets:** Normalize all active unexpired `keyset_id` strings to lowercase hexadecimal representation, and then sort them alphabetically (lexicographically by their ASCII values).

Reviewer 3: > all active unexpired `keyset_id` strings

"active" could be dropped here or the list item cleared up to include both active and inactive keysets, as active=false keysets would be included in the list.

Author: Fixed in 9af6c6d: the global commitment now explicitly includes all unexpired keysets, both active and inactive.

Verify the BIP-340 Schnorr signature `mint_signature` against the mint's master public key (`signing_pubkey` from `/v1/info`) over the SHA256 digest of the constructed epoch string.

### Step 2: Validate OpenTimestamps Attestation

1. **Verify or Upgrade Receipt:** The verifier retrieves the `ots_receipt`. If the receipt contains a pending tag (`0x00 0x06`) and has remained unconfirmed/pending after a maximum timeout bound (MUST NOT be more than 24 hours from the manifest `timestamp`), the verifier MUST treat it as an **audit failure**, rather than waiting indefinitely.
2. **Upgrade Proof (if pending):** If the receipt is still pending but within the timeout window, the verifier can attempt to fetch the Merkle path by posting the `ots_receipt` to a calendar upgrade endpoint (e.g., `https://alice.btc.calendar.opentimestamps.org/upgrade`). Note, however, that a properly functioning mint carries the upgrade burden and should publish already upgraded, offline-verifiable receipts.
3. **Scan for Block:** Find block header attestation tag `0x00 0x05` (`A_BLOCKHEADER`).
4. **Parse Height:** Decode the Bitcoin block height (serialized as a VarInt) immediately following the tag.

> 5. **Confirm Block Timestamp:** Check via an independent explorer (e.g., `https://mempool.space/api/...`) that the block height exists and has sufficient confirmations. Furthermore, verify that the Bitcoin block's timestamp is within a reasonable tolerance window (e.g., within 4 hours) of the manifest `timestamp`, rather than requiring an exact match.

Reviewer 3: As written, Step 2: Validate OpenTimestamps Attestation doesn't verify the receipt. It finds the attestation tag, reads the block height, and checks that block exists with a plausible timestamp, but it never executes the receipt's operations (the append/prepend/sha256 chain) to confirm they end at that block's merkle root, and never checks that the receipt's starting digest is the Global Digest of the manifest being audited. Those two checks honor the OTS receipt leading from a digest to a merkle root. Skipping them means a mint can hand-write a receipt pointing at any real block from the right day and pass Step 2 without anything ever being committed on chain.

There's also a practical blocker for the starting-digest check: the Global Digest preimage covers all keysets, but the epoch's keyset set isn't published anywhere (expired keysets aren't currently discoverable), so a verifier who recomputes and gets a mismatch can't tell fraud from a missing keyset. Publishing the epoch's keyset list (or `commitment_data` itself) would fix that, and would also give Step 3.1 a defined source for "the previous epoch's global digest" as currently the only way to obtain it is parsing it out of the previous `ots_receipt`, which the spec does not currently mention.

One more: the pending tag bytes look wrong. 
Pending is `0x83dfe30d2ef90c8e` (`0x06` doesn't appear in OTS), Bitcoin's is `0x0588960d73d71901`.

Author: Fixed in 9af6c6d: manifests publish global_digest plus the ordered epoch_keysets preimage data; verification binds the receipt starting digest, executes the complete OTS operation path to the Bitcoin Merkle root, and uses the correct full pending and Bitcoin attestation tags.

### Note on Spent-Side Inflation & Keyset Rotations

A naive threat model might suggest that a malicious mint could fabricate fake spent leaves (by appending random curve points `Y` to the Spent MMR) to artificially deflate its proven outstanding liabilities. However, because the mint owns the private signing keys for its keysets, a malicious mint can always generate valid ecash signatures (`C'`) for themselves, and then "validly" spend this self-signed fake ecash. Because the mint holds the preimages/secrets for these spends, a preimage/spend verification challenge is futile and cannot prevent this behavior.

Instead, spent-side inflation is caught naturally by the **append-only nature of the MMR** and **keyset rotations/deactivations**:

1. **Append-Only Immutability:** Once a leaf is appended to either MMR, it is permanent and cannot be deleted. If the mint fabricates a spend, that spent leaf is locked into history forever.

> 2. **Keyset Rotation:** Eventually, keysets are deactivated/expired. Once deactivated, no more new ecash can be issued or spent under that keyset.

Reviewer 3: > Once deactivated, no more new ecash can be issued or spent under that keyset.

This could be a typo. Should it be "Once expired,"

Author: Fixed in 9af6c6d: deactivation now stops new issuance while allowing existing ecash to be spent until expiry; expiry stops both.

    "epoch_index": 12,
    "leaf_type": "issued | spent",
    "proof": {
      "item": "02b1a03e1b10a23429fa221087e53f19001b97ad89498a44b93b3f23a851121df4",
      "leaf_index": 45012,
      "value": 1000,
      "sibling_path": [{ "hash": "b4a1df...", "sum": 500, "is_left": true }],
      "peaks": [{ "hash": "f29a2c...", "sum": 20000 }]
    }
  }
  ```
- **Mint's Defense / Response & Verification:**
  - **If the mint is NOT in the wrong:** The mint must publish the above response containing the valid sum-MMR inclusion proof for the challenged `item` at `epoch_index` with the correct `value`.
  - **Verification:** Third parties verify that the inclusion proof walk computes up to one of the mountain peaks, and that the peaks recursively peak-bagged from right-to-left match the manifest's `issued_mmr_root_hash` / `spent_mmr_root_hash` and sum. If it verifies, the challenge is refuted.

---

### 2. History Rewriting or Append-Only Violation (`append_only_violation`)

- **Description:** The mint has modified, deleted, or reordered past leaves. This changes the internal hashes/sums of the tree, violating the strict append-only prefix preservation property.
- **Challenge Schema:**

>   The challenger presents two signed epoch manifests along with two valid inclusion proofs for the exact same `leaf_index` showing a mismatch.

Reviewer 3: Nothing in the verification walk uses `leaf_index`. A proof shows its leaf is *somewhere* in the tree, not *where*. So "two valid inclusion proofs for the exact same `leaf_index`" isn't something a third party can actually check.

That seems to allow **false convictions of an honest mint**:

1. Take a genuine proof of leaf **A** from epoch 11, and a genuine proof of a *different* leaf **B** from epoch 12. Both verify against their signed manifests.
2. Publish them as an `append_only_violation`, claiming both are index `45012`.
3. The items differ, so fraud is "proven" and the only permitted defense (showing a proof is mathematically invalid) fails, because both proofs *are* valid. The challenge can't be refuted as is.

A possible fix: the proof shape may already determine the position (`is_left` bits give the index within its mountain, peak heights give the mountain's offset), so verifiers could derive the index from the proof itself and reject any mismatches.

Author: Fixed in 9af6c6d: verifiers must now derive the leaf index from the MMR peak heights and is_left path bits, and append-only challenges are rejected unless both derived positions equal the claimed index.

- **Challenge Schema:**
  The challenger presents the two validly signed epoch manifests for `E_1` and `E_2`. The burden of proving consistency is placed entirely on the mint.
  ```json
  {
    "challenge_type": "sum_mmr_consistency_violation",
    "keyset_id": "009a6154b71113b7",
    "epoch_index_1": 11,
    "epoch_index_2": 12,
    "tree_type": "issued | spent"
  }
  ```
- **Response Schema:**
  ```json
  {
    "response_type": "sum_mmr_consistency_response",
    "keyset_id": "009a6154b71113b7",
    "epoch_index_1": 11,
    "epoch_index_2": 12,
    "tree_type": "issued | spent",
    "consistency_proof": {
      "old_size": 125000,
      "new_size": 126500,
      "proof_hashes": [{ "hash": "8f3c...", "sum": 1000000, "height": 3 }]
    }
  }
  ```
- **Mint's Defense / Response & Verification:**
  - **If the mint is NOT in the wrong:** The mint must respond to the challenge by publishing the **correct, valid MMR consistency proof** specified in the response schema (consisting of the list of sibling/peak proof hashes, sums, and heights) that mathematically merges the peaks of epoch `E_1` and the new elements to produce the peaks of epoch `E_2`.

>   - **Verification:** Third parties verify the mint's published consistency proof against the root hashes and sums in the two signed manifests. If the proof successfully validates the deterministic transition, the challenge is refuted and proven false. If the mint is silent or fails to provide a mathematically verifying consistency proof within a reasonable window, the challenge is sustained and the mint is proven fraudulent.

Reviewer 3: Practically, the mint must be made aware of the existence of any fraud challenges for it be able to respond.

Since the coordination scheme between auditor, wallet, and mint is undefined (what auditors does the mint observe regularly for challenges? etc.) perhaps it is best to leave the definition of a proven fraud up to each 3rd party auditor.

For example: Auditor X might have a precondition for fraud claims that a mint has flagged observation of it's services such that it can responsibly enforce a time window for responses.

Author: Fixed in 9af6c6d: coordination and deadlines are now explicitly auditor policy, and silence is not cryptographic proof unless the mint verifiably received the challenge and the auditor applies its published policy.

### 4. Append-Only Consistency Verification

To verify that MMR M (of size m) is a valid append-only extension of MMR N (of size n, where m > n), an auditor compares their peak lists. This append-only property can be verified in two ways:

1. **Standard MMR Consistency Proofs:** Cryptographically proving that the peaks of the old MMR N are deterministically preserved and folded into the peaks of the new MMR M.
2. **Distributed Sibling Path Prefix Checks:** Any client holding a valid inclusion proof (such as for active or spent ecash) can re-request a new proof in the following epoch and verify that the old sibling path is preserved as a direct prefix of the new sibling path. If the mint modified or removed any historical entries, the internal hashes would change, immediately violating prefix preservation.

---

## Epoch Manifests & On-Chain Commitments

Every epoch interval (e.g., 24 hours), the mint constructs and signs an Epoch Manifest:

1. **Sort Keysets:** Normalize all unexpired `keyset_id` strings (both active and inactive) to lowercase hexadecimal representation, and then sort them alphabetically (lexicographically by their ASCII values).

> 2. **Commitment Data:** Prepend the 32-byte binary `previous_global_digest` of the previous epoch (for the first epoch, this MUST be 32 bytes of zeros `0x00...00`). Then concatenate the UTF-8 lowercase `keyset_id`, 8-byte big-endian `issued_mmr_size`, 32-byte binary `issued_mmr_root_hash`, 8-byte big-endian `spent_mmr_size`, and 32-byte binary `spent_mmr_root_hash` for each keyset sequentially.

Reviewer 3: The commitment data includes each keyset's tree sizes and root hashes but not the root sums. For a size-1 tree, the root hash doesn't contain its sum since sums only enter hashing when nodes combine. For those epochs the anchor doesn't commit the sum at all; it exists only in the signed string, which isn't what gets timestamped.

This case could occur on a low-activity spent or issued tree exacerbated by short epoch intervals.

Possible fix: include `issued_mmr_root_sum` and `spent_mmr_root_sum` (8-byte big-endian) in each keyset's `commitment_data` entry, and carry them in the `epoch_keysets` entries so the preimage stays reconstructable.
